### PR TITLE
[JENKINS-59767] Replace the comma separated list for CC by a list of emails

### PR DESCRIPTION
**Pull request by aheritier**
This change is requiring a lot of refactoring (and cleanup)
[JENKINS-59767](https://issues.jenkins-ci.org/browse/JENKINS-59767)

The idea is to have this kind of result

Screenshot 2019-10-14 12 26 24

instead of 

Screenshot 2019-10-13 22 32 02

CasC is now like this (there was no release with CasC support before)

```
advisor:
  acceptToS: true
  ccs:
    - "we@email.com"
    - "they@email.com"
  email: "me@email.com"
  excludedComponents:
    - "ThreadDumps"
    - "PipelineThreadDump"
  nagDisabled: true
```

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java [hunk 1]
```diff
@@ -343,21 +379,19 @@ public String getDisplayName() {
...
+    
+    // Used from AdvisorGlobalConfiguration/index.jelly
     public FormValidation doCheckEmail(@QueryParameter String value) {
       return EmailValidator.validateEmail(value);
     }
-
-    public FormValidation doCheckCc(@QueryParameter String value) {
-      return EmailValidator.validateCC(value);
-    }
-
```
Comment (fcojfernandez) [inline src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: How do you check from UI the CC email addresses?

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java [hunk 2]
```diff
@@ -59,23 +62,33 @@
...
+   */
+  @Deprecated
+  private transient boolean isValid;
+  
   public AdvisorGlobalConfiguration() {
     load();
   }
 
-  public AdvisorGlobalConfiguration(String email, String cc, Set<String> excludedComponents) {
+  public AdvisorGlobalConfiguration(String email, List<Recipient> ccs, Set<String> excludedComponents) {
```
[the remark this answers is not in the document]
Comment (aheritier) [reply src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: Is it ? I did the test and it is working. But I replaced the @DataBoundCounstructor by @DataBoundSetter

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java [hunk 3]
```diff
@@ -254,10 +284,10 @@ public boolean selectedByDefault(Component c) {
   }
 
   public boolean isValid() {
-    return isValid(false, isAcceptToS(), getEmail(), getCc());
+    return isValid(false, isAcceptToS(), getEmail(), getCcs());
   }
 
-  public static boolean isValid(boolean logErrors, boolean acceptToS, String email, String cc) {
+  public static boolean isValid(boolean logErrors, boolean acceptToS, String email, List<Recipient> ccs) {
```
[the remark this answers is not in the document]
Comment (aheritier) [reply src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: @fcojfernandez At the java level you mean ? If another plugin was using it ?

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java (hunk 1 above)
Comment (aheritier) [reply src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: I don't it's something I forgot to remove

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java [hunk 4]
```diff
@@ -206,13 +236,13 @@ public void setEmail(@CheckForNull String email) {
     this.email = EmailUtil.fixEmptyAndTrimAllSpaces(email);
   }
 
-  public String getCc() {
-    return cc;
+  public List<Recipient> getCcs() {
```
[the remark this answers is not in the document]
Comment (aheritier) [reply src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: good idea.

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (aheritier) [reply src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: For sure @fcojfernandez I don't consider the Java compatibility but noone is depending of it. I agree that I could take care more of that

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (aheritier) [reply src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: done

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java (hunk 1 above)
Comment (aheritier) [reply src/main/java/com/cloudbees/jenkins/plugins/advisor/AdvisorGlobalConfiguration.java]: @fcojfernandez They are individually verified in the Recipient

File: src/main/java/com/cloudbees/jenkins/plugins/advisor/client/model/Recipient.java [hunk 5]
```diff
@@ -1,22 +1,50 @@
...
+  public static class DescriptorImpl extends Descriptor<Recipient> {
+    public DescriptorImpl() {
+      super.load();
+    }
+
+    public String getDisplayName() {
+      return "";
+    }
+
+    public FormValidation doCheckEmail(@QueryParameter String value) {
```
Comment (aheritier) [inline src/main/java/com/cloudbees/jenkins/plugins/advisor/client/model/Recipient.java]: @fcojfernandez CC entries are validated here

**Pull request merged**